Add option to CollectWgsMetrics to output theoretical sensitivity at various allele fractions

fleharty opened the pull request:
### Description

This adds the THEORETICAL_SENSITIVITY_OUTPUT option to CollectWgsMetrics.  This option allows a user to specify a file that will contain the theoretical sensitivities at various allele fractions.

This augments the current TheoreticalHetSensitivity so that we can now calculate theoretical sensitivities for non-germline samples.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

yfarjoun commented:
given that most of the somatic data is still exomes and not genomes, would it make sense to do this for CollectHsMetrics while you're at it?

yfarjoun commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -467,6 +470,31 @@ protected int doWork() {
+            final double[] depthDoubleArray = TheoreticalSensitivity.normalizeHistogram(collector.getUnfilteredDepthHistogram());
+            final double[] baseQDoubleArray = TheoreticalSensitivity.normalizeHistogram(collector.getUnfilteredBaseQHistogram());
+
+            collector.getUnfilteredBaseQHistogram();
```
what do these 2 lines do?

fleharty replied:
Removed these lines.  Fixed.

yfarjoun commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -467,6 +470,31 @@ protected int doWork() {
+            collector.getUnfilteredBaseQHistogram();
+            collector.getUnfilteredDepthHistogram();
+
+            TheoreticalSensitivityMetrics theoreticalSensitivityMetrics = new TheoreticalSensitivityMetrics();
```
final

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivityMetrics.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Created by fleharty on 6/30/17.
+ */
+public class TheoreticalSensitivityMetrics extends MetricBase {
```
wouldn't it be better to just emit a histogram with the values you want? it would be much easier to extend to arbitrary values. You could then also have the list of desired values as an input parameter to the CLP...

fleharty replied:
@yfarjoun I'm not entirely sure I understand what you mean by histogram.
As I understand it, the histogram stores key-value pairs where the value is a non-negative integer.  I agree with you that we should make this tool so that we can emit sensitivities for any list of allele fractions the user is interested in.

fleharty replied:
What I'm trying to say is, I don't see how to use a histogram to store a list of allele fractions with their corresponding sensitivities.

yfarjoun replied:
The "key" to the histogram is of any Comparable type. Integer is Comparable, so that works, but you can use Double as well. You could use a String even! You can see an example for having a double in the histogram here: 
picard/sam/markduplicates/util/AbstractMarkDuplicatesCommandLineProgram.java:162

fleharty replied:
fixes

yfarjoun commented:
nm. you are doing that! :-)

fleharty commented:
@yfarjoun I'd be more than happy to present on the math at the next Palantir meeting.

fleharty commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,117 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+        return new Object[][]{
+                // expected sensitivity, metrics file, allele fraction, constant depth, sample size.
+                {1.00, wgsMetricsFile, .5, 30, 10000},
+                {1.00, wgsMetricsFile, .5, 30, 10000},
```
Note to self, this is a duplicated test.

fleharty replied:
fixed.

fleharty commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,117 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+                {1.00, wgsMetricsFile, .5, 30, 10000},
+                {1.00, wgsMetricsFile, .5, 30, 10000},
+                {0.78, targetedMetricsFile, .1, 30, 10000},
+                {0.26, targetedMetricsFile, 0.1, 10, 10000}
```
I need to add a test for high depth and low allele fraction.  That was actually the original purpose of this test...

Also, I need to make sure my claim about the magic numbers coming from R is true.

fleharty commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,117 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+        final File wgsMetricsFile = new File(TEST_DIR, "test_Solexa-332667.wgs_metrics");
+
+        // This test acts primarily as an intergration test.  The sample size of 100
+        // is not quite large enough to converge properly
```
Mention that I use a small sample size so as to keep the test run time short.

fleharty replied:
fixed.

fleharty commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,117 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+        // and the depth histogram.
+
+        // We use a pretty forgiving tolerance here because for these tests
+        // we are not using large enough sample sizes to converge.
```
Again, mention that the reason here is to keep tests from running too long.

fleharty commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,117 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+    @Test (dataProvider = "equivalanceHetVsArbitrary")
+    public void testHetVsArbitrary(final File metricsFile, final double tolerance, final int sampleSize) throws Exception {
+        // This test compares Theoretical Sensitivity for arbitrary allele fractions with the theoretical het sensitivity
+        // model.  Since allele fraction of 0.5 is equivalent to a het, these should provide the same answer.
```
Mention that there are minor differences in the two models, but they shouldn't matter here.  They might matter for low allele fraction, but not for het variants.

fleharty commented:
@yfarjoun 
Thanks for the preliminary review.  
- added histogram that contains a list of allele fractions.
- added option for user specified allele fractions
- fixed issue with random seed from SamRecordSetBuilder

Back to you.

yfarjoun commented on src/main/java/picard/analysis/directed/CollectTargetedMetrics.java:
```diff
@@ -98,6 +102,11 @@ protected abstract COLLECTOR makeCollector(final Set<MetricAccumulationLevel> ac
     @Option(doc="Sample Size used for Theoretical Het Sensitivity sampling. Default is 10000.", optional = true)
     public int SAMPLE_SIZE=10000;
 
+    @Option(doc="Output for Theoretical Sensitivity metrics.  Default is null.", optional = true)
```
specify that this is for the values specified in  the `ALLELE_FRACTION` argument

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/directed/CollectTargetedMetrics.java:
```diff
@@ -98,6 +102,11 @@ protected abstract COLLECTOR makeCollector(final Set<MetricAccumulationLevel> ac
+    public File THEORETICAL_SENSITIVITY_OUTPUT;
+
+    @Option(doc="Allele fraction to run theoretical sensitivity on.", optional = true)
+    public List<Double> ALLELE_FRACTION = new LinkedList<>(Arrays.asList(0.001, 0.005, 0.01, 0.02, 0.05, 0.1, 0.2, 0.3, 0.5));
```
Arrays.asList() already returns a List<> so why do you wrap it in a LinkedList?

fleharty replied:
List<> will throw an UnsupportedOperation if  you later try to use .put().

fleharty replied:
The .put() is used in the command line parsing, so I can't easily wrap it later.

yfarjoun replied:
got it. thanks for that. so please wrap in an ArrayList instead.

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivityMetrics.java:
```diff
@@ -0,0 +1,14 @@
+ * Created by fleharty on 6/30/17.
+ */
+public class TheoreticalSensitivityMetrics extends MetricBase {
+    /** Theoretical Het Sensitivity for log-odss of 6.2 **/
```
in comment: not `Het`.

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivityMetrics.java:
```diff
@@ -0,0 +1,14 @@
+ * Created by fleharty on 6/30/17.
+ */
+public class TheoreticalSensitivityMetrics extends MetricBase {
+    /** Theoretical Het Sensitivity for log-odss of 6.2 **/
```
Do we need the fixed metrics for anything? Seems like a red-herring.

fleharty replied:
By fixed metrics, I assume you mean the TheoreticalHet sensitivities for different log odds.

I had a discussion with @davidbenjamin about the log odds threshold, and I think we need to rethink the threshold we use for TheoreticalHetSensitivity.  This was just a nice way to include THS with a log odds threshold of 6.2.

I'll remove these two because we get THS with log odds of 3.0 in the other metrics file.  And we get TS with log odds of 6.2 for 0.5 allele fraction.

fleharty replied:
Also, I was thinking about saving the 20/20/20 metric in this file.  It is a type of TheoreticalSensitivity, so it seems reasonable to me to put it there.  What do you think?

fleharty commented on src/main/java/picard/analysis/CollectRawWgsMetrics.java:
```diff
@@ -30,6 +30,8 @@
 import picard.cmdline.Option;
 import picard.cmdline.programgroups.Metrics;
 
+import java.io.File;
```
fixed

fleharty commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,162 @@ public int draw() {
+            theoreticalSensitivityMetrics.HET_SENSITIVITY_LOD6_2 = TheoreticalSensitivity.hetSNPSensitivity(depthDoubleArray, baseQDoubleArray, sampleSize, 6.2);
+            double logOddsThreshold = 6.2; // This threshold is used because it is the value used for MuTect2.
+
+            Histogram<Double> sensitivityHistogram = new Histogram<>();
```
A yes, good point.
fixed

yfarjoun commented:
Let's talk about that offline.

fleharty commented on src/main/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -467,6 +471,10 @@ protected int doWork() {
         processor.addToMetricsFile(out, INCLUDE_BQ_HISTOGRAM, dupeFilter, mapqFilter, pairFilter);
         out.write(OUTPUT);
 
+        if(THEORETICAL_SENSITIVITY_OUTPUT != null) {
```
fixed

fleharty commented:
@yfarjoun After removing the TheoreticalHetSensitivity for LOD=3 and LOD=6.2
there are no fields in the metrics file (other than the histogram).  It appears that the file MetricsFile.read() function in HTSJDK isn't set up to deal with metric files that only contain a histogram, but no other fields.

There are two solutions that I see to this.
1. Fix HTSJDK MetricsFile.read() to deal with the case where a metric file contains no fields other than the histogram.
2. Add a dummy field, to be removed after HTSJDK is fixed.

yfarjoun commented:
OK. I would say add a dummy field.

fleharty commented:
@yfarjoun Thanks again for the review, back to you.

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -41,27 +40,28 @@
     private static final Log log = Log.getInstance(TheoreticalSensitivity.class);
     private static final int SAMPLING_MAX = 600; //prevent 'infinite' loops
-    private static final int MAX_CONSIDERED_DEPTH = 1000; //no point in looking any deeper than this, otherwise GC overhead is too high.
+    private static final int MAX_CONSIDERED_DEPTH = 10000; //no point in looking any deeper than this, otherwise GC overhead is too high.
```
what if someone wants super low AF sensitivity? will the code still be able to estimate it? should this be a parameter rather than a static final?

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -143,7 +143,7 @@ We use an O(1) stochastic acceptance algorithm -- see Physica A, Volume 391, Pag
 
         RouletteWheel(final double[] weights) {
-            rng = new Random(51);
+            rng = new Random(randomSeed);
```
why do you need to control the random seed?

fleharty replied:
This was put in as one way to test convergence in one of the tests.

fleharty replied:
I suspect you are referring to the Utils in GATK4.  Unfortunately this is Picard :(

davidbenjamin replied:
Whoops, I thought this was the picard package within the GATK.

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * @param uniformRNG Random number generator to use for making draw
+     * @return Number of total successes
+     */
+    public static int binomialDraw(final int trials, final double p, final Random uniformRNG) {
```
why not use apache.Commons.BinominalDistribution?

davidbenjamin replied:
Agreed.

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     */
+    public static boolean isCalled(int totalDepth, int altDepth, double averageQuality, double alleleFraction, double logOddsThreshold) {
+        double threshold;
+        double sumOfQualities = altDepth * averageQuality;
```
is there a reason you want to pass in averageQuality and not sumOfQualities? that seems to be the original value and also the one you need...so why divide and then multiply?

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+    public static boolean isCalled(int totalDepth, int altDepth, double averageQuality, double alleleFraction, double logOddsThreshold) {
+        double threshold;
+        double sumOfQualities = altDepth * averageQuality;
+        threshold = 10.0 * (altDepth * Math.log10(1.0 / alleleFraction) + (totalDepth - altDepth) * Math.log10(1.0 / (1.0 - alleleFraction)) + logOddsThreshold);
```
instead of the 1/x you can just put a `-` in front of the log10

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+            throw new PicardException("Allele fractions must be between 0 and 1.");
+        }
+        double sensitivity = 0.0;
+        for (int k = 0; k < depthDistribution.length; k++) {
```
if `k` is depth, then call it `depth`, not `k`.

yfarjoun replied:
Given that you want to go to high depth, it would make sense to bin the depth distribution in a smart manner (that gives higher fidelity for small depths) and iterate over that binned distribution.

fleharty replied:
We discussed using a log distribution on depth for each sensitivityAtConstantDepth.  I think this is a sensible thing to do, but I think there are other (hopefully even better) ways to speed this up that I will discuss with @davidbenjamin

davidbenjamin replied:
Let's discuss these. . .

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+
+    /**
+     * This is a utility function
+     * @param theoreticalSensitivityOutput File to save to results ot theoretical sensitivity.
```
re-read/re-write javadoc

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * This is a utility function
+     * @param theoreticalSensitivityOutput File to save to results ot theoretical sensitivity.
+     * @param tsOut MetricsFile object to save results of theoretical sensitivity to.
+     * @param sampleSize Number of samples to take for each depth.
```
the word `sample` is used in two different meaning in this javadoc...needs clarification

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * @param sampleSize Number of samples to take for each depth.
+     * @param depthHistogram Histogram of depth distribution for sample.
+     * @param baseQHistogram Histogram of Phred-scaled quality scores.
+     * @param alleleFractions Allele fractions
```
alleleFractions needs more detail on what the parameter is.

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivityMetrics.java:
```diff
@@ -0,0 +1,11 @@
+public class TheoreticalSensitivityMetrics extends MetricBase {
+    /** This is a dummy value that allows for this to be read properly from a file. */
+    public int DUMMY_FIELD = 0;
+}
```
another newline in the end to make github happy

fleharty replied:
fixed

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+
+        int altDepth = 0;
+        int calledVariants = 0;
+        for (int k = 0; k < sampleSize; k++) {
```
using some properties of the qDist, (mean and stdev) you should be able to figure out at what point (altDepth+alleleFraction) you can ignore the details of the distribution and use the meanQuality directly to approximate sumOfQualities

fleharty replied:
That's true, but this would be an additional approximation to speed things up when the runtime isn't very long now.  I think if we find speed to be an issue I'd be happy to revisit this.

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+
+        // Locate the index of the distribution where all the values remaining at
+        // larger indices are zero.
+        for (endOfDistribution = distribution.length-1;endOfDistribution >= 0;endOfDistribution--) {
```
spaces around `;` and `-`

yfarjoun commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+
+        // Remove trailing zeros.
+        final double[] trimmedDistribution = new double[endOfDistribution+1];
+        for (int i = 0;i <= endOfDistribution;i++) {
```
spaces

fleharty replied:
fixed

yfarjoun commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,118 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+        // is not quite large enough to converge properly, but is used for the purpose of
+        // keeping the compute time of the tests short.
+        return new Object[][] {
+                {0.90, wgsMetricsFile,       .5, 100},
```
why so much space here?

fleharty replied:
fixed

fleharty commented:
@davidbenjamin Thanks for agreeing to review this.
If you would like for me to rebase or anything to make it easier for you to review let me know.

Also, I anticipate an objection from you on the issue of how it converges, it's super slow.  I would like to speed it up, and if you have thoughts on how best to do that, I would be open to making those changes.

fleharty commented:
@davidbenjamin 
Before you review, I'll address Yossi's comments, and then I'll send it back to you.

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * @param logOddsThreshold Log odds threshold necessary to exceed for variant to be called
+     * @return
+     */
+    public static boolean isCalled(int totalDepth, int altDepth, double averageQuality, double alleleFraction, double logOddsThreshold) {
```
What, in words, is the criterion for `isCalled`?  I'm asking because `allele fraction` here seems possibly different from the command line argument allele fraction.  The latter, I think, is the true allele fraction, which the former can't be because only the alt and total depths, which are subject to sampling error, are observed.

I could imagine various ways to define theoretical sensitivity with the allele fraction in this method being:

* an exact value.  `isCalled` is true if the likelihood ratio between a) het with exactly this allele fraction and b) hom ref  is sufficiently large.
* a minimum. `isCalled` is true if the likelihood of being a het with *at least* this allele fraction is sufficiently large.
* irrelevant. `isCalled` could use a criterion that marginalizes out the latent true allele fraction and does model comparison between a) hom ref and b) int_0^1 likelihood(af) P(af) d af.  This is what Mutect now does.
* irrelevant. `isCalled` plugs in af = alt / total in calculating the likelihood.  This is what Mutect used to do.

I'm not saying any one is more right than the others, (well, I guess that between the last two my preference is clear) but I want to know which one the code implements.

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * depth.
+     * @param depth Depth to compute sensitivity at
+     * @param qualityDistribution Phred-scaled quality score distribution
+     * @param logOddsThreshold Log odd threshold necessary to exceed for variant to be called
```
Log odd*s*

fleharty replied:
fixed

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * depth.
+     * @param depth Depth to compute sensitivity at
+     * @param qualityDistribution Phred-scaled quality score distribution
+     * @param logOddsThreshold Log odd threshold necessary to exceed for variant to be called
```
And same as above about "necessary to exceed"

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     */
+    public static double sensitivityAtConstantDepth(final int depth, final double[] qualityDistribution, final double logOddsThreshold, final int sampleSize, final double alleleFraction, final int randomSeed) {
+        final RouletteWheel qualityRW = new RouletteWheel(trimDistribution(qualityDistribution));
+        final Random uniformRNG = new Random(randomSeed);
```
Utils.getRandomGenerator

davidbenjamin replied:
If Picard has such a thing.

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+        final RouletteWheel qualityRW = new RouletteWheel(trimDistribution(qualityDistribution));
+        final Random uniformRNG = new Random(randomSeed);
+
+        int altDepth = 0;
```
This variable is local to the `for` loop and should not be declared outside it.

fleharty replied:
fixed

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+        int altDepth = 0;
+        int calledVariants = 0;
+        for (int k = 0; k < sampleSize; k++) {
+            altDepth = binomialDraw(depth, alleleFraction, uniformRNG);
```
`final`

fleharty replied:
fixed

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+        for (int k = 0; k < sampleSize; k++) {
+            altDepth = binomialDraw(depth, alleleFraction, uniformRNG);
+
+            int sumOfQualities = 0;
```
`final int sumOfQualities = IntStream.range(0, altDepth).map(n -> qualityRW.draw()).sum()`

fleharty replied:
fixed

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     */
+    public static double theoreticalSensitivity(final double[] depthDistribution, final double[] qualityDistribution,
+                                                final int sampleSize, final double logOddsThreshold, final double alleleFraction) {
+        if (alleleFraction > 1.0 || alleleFraction < 0.0) {
```
`Utils.validateArg()`

davidbenjamin replied:
If Picard has such a thing.

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+
+        // Remove trailing zeros.
+        final double[] trimmedDistribution = new double[endOfDistribution+1];
+        for (int i = 0;i <= endOfDistribution;i++) {
```
Use `Arrays.copyOfRange()`

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * This is a utility function
+     * @param theoreticalSensitivityOutput File to save to results ot theoretical sensitivity.
+     * @param tsOut MetricsFile object to save results of theoretical sensitivity to.
+     * @param sampleSize Number of samples to take for each depth.
```
Elsewhere you use `simulations`, which would seem to work here, too.

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     */
+    public static void writeOutput(final File theoreticalSensitivityOutput, final MetricsFile<TheoreticalSensitivityMetrics, Double> tsOut, final int sampleSize,
+                                   final Histogram depthHistogram, final Histogram baseQHistogram, final List<Double> alleleFractions) {
+        if (theoreticalSensitivityOutput != null) {
```
To reduce nesting, use 
```java
if (theoreticalSensitivityOutput == null) {
    return;
}
```

fleharty replied:
fixed

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+    public static void writeOutput(final File theoreticalSensitivityOutput, final MetricsFile<TheoreticalSensitivityMetrics, Double> tsOut, final int sampleSize,
+                                   final Histogram depthHistogram, final Histogram baseQHistogram, final List<Double> alleleFractions) {
+        if (theoreticalSensitivityOutput != null) {
+            final double logOddsThreshold = 6.2; // This threshold is used because it is the value used for MuTect2.
```
But (see above) is the criterion for making a call the same as Mutect's?

fleharty replied:
Honestly, I don't know, and I don't know to what you are referring to "see above"

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+                                   final Histogram depthHistogram, final Histogram baseQHistogram, final List<Double> alleleFractions) {
+        if (theoreticalSensitivityOutput != null) {
+            final double logOddsThreshold = 6.2; // This threshold is used because it is the value used for MuTect2.
+            final double[] depthDoubleArray = TheoreticalSensitivity.normalizeHistogram(depthHistogram);
```
`depthDistribution` or `depthPdf` would be a more descriptive name.

fleharty replied:
fixed.

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+        if (theoreticalSensitivityOutput != null) {
+            final double logOddsThreshold = 6.2; // This threshold is used because it is the value used for MuTect2.
+            final double[] depthDoubleArray = TheoreticalSensitivity.normalizeHistogram(depthHistogram);
+            final double[] baseQDoubleArray = TheoreticalSensitivity.normalizeHistogram(baseQHistogram);
```
Likewise.  `baseQualityDistribution`.

fleharty replied:
fixed

davidbenjamin commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+                                   final Histogram depthHistogram, final Histogram baseQHistogram, final List<Double> alleleFractions) {
+        if (theoreticalSensitivityOutput != null) {
+            final double logOddsThreshold = 6.2; // This threshold is used because it is the value used for MuTect2.
+            final double[] depthDoubleArray = TheoreticalSensitivity.normalizeHistogram(depthHistogram);
```
Is `normalizeHistogram` any different from `MathUtils.normalizeFromRealSpace`?

fleharty replied:
I can't find this function normalizeFromRealSpace.

davidbenjamin commented on src/test/java/picard/analysis/TheoreticalSensitivityTest.java:
```diff
@@ -216,13 +216,118 @@ public void testHetSensTargeted(final double expected, final File metricsFile) t
+        final double resultFromTHS = TheoreticalSensitivity.hetSNPSensitivity(depthDistribution, qualityDistribution, sampleSize, 3);
+
+        Assert.assertEquals(resultFromTHS, resultFromTS, tolerance);
+    }
```
I would like to see one unit test with a back-of-the-envelope result.

yfarjoun commented:
this PR needs a rebase and please respond to review comments....

fleharty commented:
@yfarjoun Noted.  The algorithm as it is is too slow.  I need to make a few improvements to speed it up.

yfarjoun commented:
bump!

fleharty commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     */
+    public static boolean isCalled(int totalDepth, int altDepth, double averageQuality, double alleleFraction, double logOddsThreshold) {
+        double threshold;
+        double sumOfQualities = altDepth * averageQuality;
```
fixed

fleharty commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * @return
+     */
+    public static boolean isCalled(int totalDepth, int altDepth, double averageQuality, double alleleFraction, double logOddsThreshold) {
+        double threshold;
```
fixed

fleharty commented on src/main/java/picard/analysis/TheoreticalSensitivity.java:
```diff
@@ -204,4 +204,174 @@ public int draw() {
+     * @param logOddsThreshold Log odds threshold necessary to exceed for variant to be called
+     * @return
+     */
+    public static boolean isCalled(int totalDepth, int altDepth, double averageQuality, double alleleFraction, double logOddsThreshold) {
```
fixed
